### PR TITLE
Set PROGRAM_FILE global via `argv[0]`.

### LINKED ISSUE #54: `PROGRAM_FILE` not set when compiled through `juliac.jl`
`Base.PROGRAM_FILE` is empty when compiled through `juliac.jl`. (The global `ARGS` is empty, which is where `PROGRAM_FILE` is set from.)

`PROGRAM_FILE` is set [here](https://github.com/JuliaLang/julia/blob/v0.6.2/base/client.jl#L301), by popping the first value off of `ARGS`, iff `ARGS` is not empty (checked [here](https://github.com/JuliaLang/julia/blob/33fdd3/base/client.jl#L271)).

Currently, the default `program.c` skips setting the global `ARGS`, opting instead to [pass an Array to `julia_main`](https://github.com/JuliaLang/PackageCompiler.jl/blob/b8f0d0a/examples/program.c#L32-L40). This means that there is no way to access `PROGRAM_FILE` from compiled code.

### PR DESCRIPTION
**Pull request by NHDaly**
Changes default `program.c` to manually set `PROGRAM_FILE` via its `argv[0]`.

Ideally, I think we would also set global `ARGS`, in the same way the julia interpreter does, but I still haven't figured out how to get that to work.

For #54.

### CODE REVIEW

**Comment by NHDaly** [comment]
It's worth mentioning that in normal julia programs, `PROGRAM_FILE` is set by popping the first arg of the global `ARGS`:
https://github.com/JuliaLang/julia/blob/33fdd3/base/client.jl#L289

I tried just calling [`jl_set_ARGS(argc, argv)`](https://github.com/JuliaLang/julia/blob/v0.6.2/src/jloptions.c#L516), but no matter where I put it in `program.c`, the global `ARGS` was never set. Perhaps this is a bug in julia?

**Comment by lucatrv** [comment]
Before merging this PR, maybe you should verify if it is actually due to a Julia bug, either on Discourse, or otherwise directly opening an issue on Julia's repository if you believe it is a bug. What do you think?

**Comment by ihnorton** [comment]
Note that the assignment block in `jl_set_ARGS` is only executed if `jl_core_module` is initialized (`jl_core_module != NULL`). I don't know much about the structure of this package, but from a quick glance at program.c, I don't see any explicit initialization of that variable.

**Comment by NHDaly** [comment]
Hmm, yeah i saw that, but it seems that yes, `jl_core_module` is initialized. I walked through with `lldb`, and it doesn't take the branch, since `jl_core_module` has a value, and so it does in fact execute the assignment block. Sorry, i should've mentioned that earlier.
In fact, `args` is not `null`, and it calls `jl_arrayset` for each argument supplied. Which is why it baffled me that there still isn't anything in the julia `ARGS` afterwards.

For completeness, i think the core_module is initialized here:
https://github.com/JuliaLang/julia/blob/master/src/staticdata.c#L1511

**Comment by NHDaly** [comment]
I'll ask around on slack and/or file a bug. thanks for the help!

**Comment by NHDaly** [comment]
Okay, so i've found the problem! It was not a bug, it turns out, but rather it's built-in behavior that the REPL does, which we weren't replicating.

Basically, we were reading from `Base.ARGS` (global `ARGS` points there). But `jl_set_ARGS` only sets the _Core.ARGS_!

In a normal program, the REPL automatically copies `Core.ARGS` into `Base.ARGS`, but we weren't getting that behavior.

ARGS are set here:
https://github.com/JuliaLang/julia/blob/v0.6.3/base/client.jl#L367-L368
and the PROGRAM_FILE is set here:
https://github.com/JuliaLang/julia/blob/v0.6.3/base/client.jl#L301

But `_start()` is only called from the repl, here:
https://github.com/JuliaLang/julia/blob/v0.6.3/ui/repl.c#L121


--------------

So we just need to either replicate that behavior from `program.c`, or petition Julia to set both Base.ARGS and Core.ARGS in `jl_set_ARGS`, or both.

File: examples/program.c [hunk 1]
```diff
@@ -36,6 +44,37 @@ int main(int argc, char *argv[])
...
+    // Set Base.ARGS to argv[1:end], to match behavior of Base._start().
+    //  -- jl_set_const doesn't work because of https://github.com/JuliaLang/julia/pull/27520.
+    //jl_set_const(jl_base_module, jl_symbol("ARGS"), (jl_value_t*)ARGS);
+    //  -- This should work, but for some reason it doesn't. It says it's
+    //  -- "WARNING: redefining constant ARGS", but for w/e reason doesn't...
+    //  -- I don't understand why it's not working.
+    //jl_checked_assignment(jl_get_binding_wr(jl_base_module, jl_symbol("ARGS")),
+    //                       (jl_value_t*)ARGS);
+    // -- Trying this instead, even though it's bulky:
+    set_julia_base_ARGS(argc, argv);
```
Comment (NHDaly) [inline examples/program.c]: So all of the above is really wonky... I first just tried overwriting the Base.ARGS via `jl_set_const`, but ran into this problem:
https://github.com/JuliaLang/julia/pull/27520

So then I tried overwriting it via `jl_checked_assignment`, which really should work, but for some reason the compiled program still has an empty `String[]` for ARGS, and i don't know why...

So finally, i've caved and i'm just manually appending all the fields to the Base.ARGS, just like `Base._start()` does, via this helper function `set_julia_base_ARGS`.

**Comment by NHDaly** [comment]
(And we should squash at least the first two commits, before merging; i only left the first one there for review clarity. 😊)

**Comment by lucatrv** [comment]
CI tests failing on Windows...

**Comment by NHDaly** [comment]
Oops, thanks, I forgot about the `executable_ext` for windows. Repushing now...

Also, I took this chance to squash & reword the commits. :)

**Comment by lucatrv** [comment]
Still failing...

**Comment by NHDaly** [comment]
gah, sorry. i don't have access to a windows machine so it's hard for me to test... i'll see what i can do...

**Comment by NHDaly** [comment]
Okay finally fixed!! I left the println in the test because actually I think it's useful overall.

Thanks for reminding me when it was failing! :)

**Comment by lucatrv** [comment]
Merged!

**Comment by NHDaly** [comment]
:) thanks!

**Pull request merged**